### PR TITLE
pacman module: Improve query performance

### LINKED ISSUE #31223: Improve pacman module performance
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
module:pacman

##### ANSIBLE VERSION
```
ansible 2.4.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/jonas/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.14 (default, Sep 20 2017, 01:25:59) [GCC 7.2.0]
```

When using `state: present`, the `pacman` module still queries the local and remote version of each package instead of doing a simple `pacman -Q <package>`, which is wasteful and really pushes on the performance for lower spec devices.

https://github.com/ansible/ansible/blob/ef25367353b3b01813a8b2681c3f9b08227fc5e9/lib/ansible/modules/packaging/os/pacman.py#L143

This is what the command effectively executes in this situation:

```
# time bash -c 'pacman -Qi bash && pacman -Si bash'
Name            : bash
Version         : 4.4.012-2
Description     : The GNU Bourne Again shell
Architecture    : armv6h
URL             : http://www.gnu.org/software/bash/bash.html
Licenses        : GPL
Groups          : base
Provides        : sh
Depends On      : readline>=7.0  glibc  ncurses
Optional Deps   : bash-completion: for tab completion
Required By     : autoconf  automake  bison  bzip2  ca-certificates-utils  db  dhcpcd  dialog
                  diffutils  e2fsprogs  fakeroot  findutils  flex  gawk  gdbm  gettext  gmp  gpm  gzip
                  icu  iptables  keyutils  libgpg-error  libksba  libpcap  libtool  lvm2  m4  man-db
                  mkinitcpio  nano  openresolv  pacman  pcre  pcre2  shadow  systemd  texinfo  which
                  wireless-regdb  xfsprogs  xz
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 7.31 MiB
Packager        : Arch Linux ARM Build System <builder+xu8@archlinuxarm.org>
Build Date      : Tue Feb 21 19:18:47 2017
Install Date    : Sun Sep 10 16:32:21 2017
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : SHA-256 Sum

Repository      : core
Name            : bash
Version         : 4.4.012-2
Description     : The GNU Bourne Again shell
Architecture    : armv6h
URL             : http://www.gnu.org/software/bash/bash.html
Licenses        : GPL
Groups          : base
Provides        : sh
Depends On      : readline>=7.0  glibc  ncurses
Optional Deps   : bash-completion: for tab completion
Conflicts With  : None
Replaces        : None
Download Size   : 1346.20 KiB
Installed Size  : 7489.00 KiB
Packager        : Arch Linux ARM Build System <builder+xu8@archlinuxarm.org>
Build Date      : Tue Feb 21 19:18:47 2017
Validated By    : MD5 Sum  SHA-256 Sum  Signature


real	0m10.390s
user	0m4.820s
sys	0m0.310s
```

The much easier alternative would be:

```
# time pacman -Q bash 
bash 4.4.012-2

real	0m0.220s
user	0m0.090s
sys	0m0.020s
```

15.5s vs. 0.33s, for every installed package!

Am I missing something here?

### PR DESCRIPTION
**Pull request by MazeChaZer**
Fixes #31223

##### SUMMARY

Previously, the pacman module would always run `pacman -Qi <package>` followed by `pacman -Si <package>`, even if the extended information from these commands aren't needed. This would considerably slow down the execution.
For me, this fix lowered the execution time of my playbook from 20m to 8m, both without any changes applied.
The behaviour of the module has not been changed and should be 100% backward-compatible.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
module:pacman

##### ANSIBLE VERSION
```
ansible 2.4.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/jonas/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.14 (default, Sep 20 2017, 01:25:59) [GCC 7.2.0]
```

### CODE REVIEW

**Comment by elasticdog** [comment]
Why does @ansibot still ping me after https://github.com/ansible/ansible/pull/32051 was merged??? I just checked the devel branch and my username is no longer anywhere in the _[BOTMETA.yml](https://github.com/ansible/ansible/blob/devel/.github/BOTMETA.yml)_ file.

**Comment by MazeChaZer** [comment]
Hey @resmo, I know it's been a while since I wrote on ansible-devel about this issue, but if you are still up for a review I would appreciate it

File: lib/ansible/modules/packaging/os/pacman.py [hunk 1]
```diff
@@ -335,11 +347,15 @@ def check_packages(module, pacman_path, packages, state):
...
 
     for package in packages:
-        installed, updated, unknown = query_package(module, pacman_path, package)
-        if ((state in ["present", "latest"] and not installed) or
-                (state == "absent" and installed) or
-                (state == "latest" and not updated)):
-            would_be_changed.append(package)
+        if state == "latest":
+            installed, updated, error = query_package_latest(module, pacman_path, package)
+            if not updated:
```
Comment (tchernomax) [inline lib/ansible/modules/packaging/os/pacman.py]: updated→uptodate
It's easier to understand.

File: lib/ansible/modules/packaging/os/pacman.py [hunk 2]
```diff
@@ -271,13 +279,17 @@ def install_packages(module, pacman_path, state, packages, package_files):
...
+            if installed:
+                continue
+        elif state == "latest":
+            installed, updated, error = query_package_latest(module, pacman_path, package)
+            if error:
+                package_err.append(package)
 
-        if installed and (state == 'present' or (state == 'latest' and updated)):
-            continue
+            if updated:
```
Comment (tchernomax) [inline lib/ansible/modules/packaging/os/pacman.py]: updated→uptodate (or up_to_date)

File: lib/ansible/modules/packaging/os/pacman.py [hunk 3]
```diff
@@ -138,32 +138,40 @@ def get_version(pacman_output):
     return None
 
 
-def query_package(module, pacman_path, name, state="present"):
-    """Query the package status in both the local system and the repository. Returns a boolean to indicate if the package is installed, a second
-    boolean to indicate if the package is up-to-date and a third boolean to indicate whether online information were available
+def query_package_installed(module, pacman_path, name):
```
Comment (tchernomax) [inline lib/ansible/modules/packaging/os/pacman.py]: What if we implement query_package_local and query_package_remote instead ?
(replacing query_package_installed and query_package_latest)
What do you think ? (just a suggestion)

**Review by tchernomax — COMMENTED** [review]
I would like to see this PR merged ! Could you rebase it and think about the comments I made ?
Regards

**Comment by MazeChaZer** [comment]
@tchernomax Great! It's been quite some time since I submitted this ^^
Will do a rebase and look over your comments next week :+1:

**Comment by tchernomax** [comment]
@MazeChaZer don't get me wrong, a PR of mine have been merged recently regarding pacman. Since then I start to receive every notifications (issues and PRs) regarding this module. This PR was part of the bulk.
However, if I flag this PR for merging it still may not be enough for the ansibot to accept it (we will see).

**Comment by MazeChaZer** [comment]
I just rebased this branch against the latest `devel` and considered your feedback, @tchernomax

**Review by tchernomax — APPROVED** [review]
LGTM

**Pull request closed without merging**